libraries/pod/src/optional_keys.rs

```diff
@@ -0,0 +1,340 @@
+struct OptionalNonZeroPubkeyVisitor;
+
+#[cfg(feature = "serde-traits")]
+impl<'de> Visitor<'de> for OptionalNonZeroPubkeyVisitor {
+    type Value = OptionalNonZeroPubkey;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a Pubkey in base58 or `null`")
+    }
+
+    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+    where
+        E: Error,
+    {
+        let pkey = Pubkey::from_str(&v)
+            .map_err(|_| Error::invalid_value(Unexpected::Str(v), &"value string"))?;
+
+        OptionalNonZeroPubkey::try_from(Some(pkey))
+            .map_err(|_| Error::custom("Failed to convert from pubkey"))
+    }
+
+    fn visit_unit<E>(self) -> Result<Self::Value, E>
+    where
+        E: Error,
+    {
+        OptionalNonZeroPubkey::try_from(None).map_err(|e| Error::custom(e.to_string()))
+    }
+}
+
+#[cfg(feature = "serde-traits")]
+impl<'de> Deserialize<'de> for OptionalNonZeroPubkey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_any(OptionalNonZeroPubkeyVisitor)
+    }
+}
+
+/// An ElGamalPubkey that encodes `None` as all `0`, meant to be usable as a Pod type.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Pod, Zeroable)]
+#[repr(transparent)]
+pub struct OptionalNonZeroElGamalPubkey(ElGamalPubkey);
+impl OptionalNonZeroElGamalPubkey {
+    /// Checks equality between an OptionalNonZeroElGamalPubkey and an ElGamalPubkey when
+    /// interpreted as bytes.
+    pub fn equals(&self, other: &ElGamalPubkey) -> bool {
+        &self.0 == other
+    }
+}
+impl TryFrom<Option<ElGamalPubkey>> for OptionalNonZeroElGamalPubkey {
+    type Error = ProgramError;
+    fn try_from(p: Option<ElGamalPubkey>) -> Result<Self, Self::Error> {
+        match p {
+            None => Ok(Self(ElGamalPubkey::default())),
+            Some(elgamal_pubkey) => {
+                if elgamal_pubkey == ElGamalPubkey::default() {
+                    Err(ProgramError::InvalidArgument)
+                } else {
+                    Ok(Self(elgamal_pubkey))
+                }
+            }
+        }
+    }
+}
+impl From<OptionalNonZeroElGamalPubkey> for Option<ElGamalPubkey> {
+    fn from(p: OptionalNonZeroElGamalPubkey) -> Self {
+        if p.0 == ElGamalPubkey::default() {
+            None
+        } else {
+            Some(p.0)
+        }
+    }
+}
+
+#[cfg(any(feature = "serde-traits", test))]
+const OPTIONAL_NONZERO_ELGAMAL_PUBKEY_LEN: usize = 32;
+
+#[cfg(feature = "serde-traits")]
+impl Serialize for OptionalNonZeroElGamalPubkey {
+    fn serialize<S>(&self, s: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if self.0 == ElGamalPubkey::default() {
+            s.serialize_none()
+        } else {
+            s.serialize_some(&self.0.to_string())
+        }
+    }
+}
+
+#[cfg(feature = "serde-traits")]
+pub(crate) struct OptionalNonZeroElGamalPubkeyVisitor;
```

Comment:
micro-nit: does this need to be `pub` at all? seems like it's only used in the `Serialize` / `Deserialize` impls